presto-main/src/main/java/io/prestosql/operator/NestedLoopJoinPagesBuilder.java

```diff
@@ -47,12 +51,44 @@ public void addPage(Page page)
             return;
         }
 
+        // Fast path for empty output channels
+        if (page.getChannelCount() == 0) {
+            updatePagePositionCounter(page.getPositionCount());
+            return;
+        }
+
         pages.add(page);
         estimatedSize += page.getRetainedSizeInBytes();
     }
 
+    private void updatePagePositionCounter(int positions)
+    {
+        try {
+            this.emptyChannelPositionCounter = addExact(this.emptyChannelPositionCounter, positions);
+        }
+        catch (ArithmeticException overflow) {
+            flushPositionCounter();
+            this.emptyChannelPositionCounter = positions;
+        }
+    }
+
+    private void flushPositionCounter()
+    {
+        checkState(!finished, "NestedLoopJoinPagesBuilder is finished");
+        if (emptyChannelPositionCounter > 0) {
+            Page output = new Page(emptyChannelPositionCounter);
+            pages.add(output);
+            estimatedSize += output.getRetainedSizeInBytes();
+            this.emptyChannelPositionCounter = 0;
+        }
+    }
+
     public DataSize getEstimatedSize()
     {
+        if (emptyChannelPositionCounter > 0) {
+            // account for pending positionCount only page
+            return DataSize.ofBytes(estimatedSize + POSITION_ONLY_PAGE_RETAINED_SIZE);
```

Comment:
Not sure I understand. Where is the "pending" page being held? From my reading of the code above, there's no page involved. It's just an int in this  class until it's time to flush it, at which point a page is created and estimatedSize is updated.

Reply:
That's true. I wanted to avoid suddenly having an estimated size that increases right as the operator completes (even though the growth here is supposed to in at most, a few bytes). Basically, accounting for the page size that we know will be created when the build side finishes as soon as we know that it will be created. I'm fine removing it if you think it's pointless.

Reply:
It seems overkill, so I'd remove it to keep the code simpler.

Reply:
Removed